metadata/src/main/java/org/apache/kafka/metadata/bootstrap/BootstrapMetadata.java

```diff
@@ -73,6 +80,22 @@ public static Optional<MetadataVersion> recordToMetadataVersion(ApiMessage recor
         return Optional.empty();
     }
 
+    private static ApiMessageAndVersion kraftVersionToRecord(KRaftVersion version) {
+        return new ApiMessageAndVersion(new FeatureLevelRecord().
+            setName(KRaftVersion.FEATURE_NAME).
+            setFeatureLevel(version.featureLevel()), (short) 0);
+    }
+
+    private static Optional<KRaftVersion> recordToKRaftVersion(ApiMessage record) {
+        if (record instanceof FeatureLevelRecord) {
+            FeatureLevelRecord featureLevel = (FeatureLevelRecord) record;
+            if (featureLevel.name().equals(KRaftVersion.FEATURE_NAME)) {
+                return Optional.of(KRaftVersion.fromFeatureLevel(featureLevel.featureLevel()));
+            }
+        }
+        return Optional.empty();
+    }
```

Comment:
I don't think we should do this. The kraft.version is written to the cluster metadata log using the KRaftVersion control record. These are the cases when KRaft will write this control record:

1. The kafka-storage tool will write it to the checkpoint/snapshot if it being formatted to support kraft.version 1. [Storage tool changes for KIP-853](https://issues.apache.org/jira/browse/KAFKA-16518)
2. `KafkaRaftClient` will write the KRaftVersion control record to the checkpoint/snapshot using `RecordsSnapshotWriter.Builder` if the control record is part of the log segments. This is already in trunk.
3. `KafkaRaftClient` will write the `KRaftVersion` control record when handling `UpdateFeatures`. [UpdateFeatures for kraft.version](https://issues.apache.org/jira/browse/KAFKA-16538)
4. The first KRaft leader will write the `KRaftVersion` control record if it is in the bootstrapping checkpoint/snapshot but it is has never been written to the log segments. [Support for first leader bootstrapping the voter set](https://issues.apache.org/jira/browse/KAFKA-16532)

I think that KRaftVersion should behave different from other feature level in that the KRaft version should be persisted and replicated using control records instead of metadata records. This is needed mainly because KRaft needs to read and process this record as uncommitted data. The controller and the metadata layer only sees committed records.

Reply:
ack